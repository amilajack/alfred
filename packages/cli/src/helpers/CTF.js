import path from 'path';
import fs from 'fs';
import npm from 'npm';
import yarn from 'yarn-api';
import CTF, {
  getDevDependencies,
  CORE_CTFS,
  INTERFACE_STATES,
  normalizeInterfacesOfSkill,
  AddCtfHelpers
} from '@alfredpkg/core';
import type { CtfMap, InterfaceState } from '@alfredpkg/core';
import ValidateConfig from './Validation';
import { getProjectRoot } from './CLI';

export const ENTRYPOINTS = [
  'lib.node.js',
  'app.node.js',
  'lib.browser.js',
  'app.browser.js',
  'lib.electron.main.js',
  'lib.electron.renderer.js',
  'app.electron.main.js',
  'app.electron.renderer.js',
  'lib.react-native.js',
  'app.react-native.js'
];

const projectRoot = getProjectRoot();

export function generateInterfaceStatesFromProject(): Array<InterfaceState> {
  const envs = ['production', 'development', 'test'];
  // Default to development env if no config given
  const env = envs.includes(process.env.NODE_ENV)
    ? process.env.NODE_ENV
    : 'development';

  return ENTRYPOINTS.filter(e =>
    fs.existsSync(path.join(projectRoot, 'src', e))
  ).map(e => {
    const [projectType, target] = e.split('.');
    return {
      env,
      target,
      projectType
    };
  });
}

/**
 * Find all the dependencies that are different between two CTF's.
 * This is used to figure out which deps need to be installed
 */
export function diffCtfDeps(oldCtf: CtfMap, newCtf: CtfMap): Array<string> {
  // Find the dependencies that have changed and install them
  const t: Map<string, string> = new Map();
  const s: Map<string, string> = new Map();

  Object.entries(getDevDependencies(oldCtf)).forEach(([key, val]) => {
    t.set(key, val);
  });
  Object.entries(getDevDependencies(newCtf)).forEach(([key, val]) => {
    if (t.has(key)) {
      if (t.get(key) !== val) {
        throw new Error('Cannot resolve diff deps');
      }
    } else {
      s.set(key, val);
    }
  });

  return Array.from(s.entries()).map(([key, val]) => `${key}@${val}`);
}

/**
 * @TODO Account for `devDependencies` and `dependencies`
 */
export function installDeps(
  dependencies: Array<string> = [],
  npmClient: 'npm' | 'yarn' = 'npm'
): Promise<any> {
  if (!dependencies.length) return Promise.resolve();

  switch (npmClient) {
    case 'npm': {
      return new Promise((resolve, reject) => {
        npm.load({ save: true }, err => {
          if (err) reject(err);

          npm.commands.install(dependencies, (_err, data) => {
            if (_err) reject(_err);
            resolve(data);
          });

          npm.on('log', console.log);
        });
      });
    }
    case 'yarn': {
      return new Promise((resolve, reject) => {
        yarn(['why', 'isobject'], err => {
          if (err) reject(err);
          resolve();
        });
      });
    }
    default: {
      throw new Error('Unsupported npm client. Can only be "npm" or "yarn"');
    }
  }
}

export type AlfredConfig = {
  npmClient: 'npm' | 'yarn',
  skills: Array<string>,
  root: string,
  showConfigs: boolean
};

/**
 * Add skills to a given list of skills to ensure that the list has a complete set
 * of standard ctfs
 */
export function addMissingStdSkillsToCtf(
  ctf: CtfMap,
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): CtfMap {
  const stdCtf = new Map(
    Object.entries({
      lint: CORE_CTFS.eslint,
      format: CORE_CTFS.prettier,
      build: require('@alfredpkg/interface-build').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      start: require('@alfredpkg/interface-start').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      test: CORE_CTFS.jest
    })
  );

  ctf.forEach(ctfNode => {
    /* eslint no-param-reassign: off */
    ctfNode.interfaces = normalizeInterfacesOfSkill(ctfNode.interfaces);
  });

  const stdSubCommands: Set<string> = new Set(stdCtf.keys());
  // Create a set of subcommands that the given CTF has
  const ctfSubcommands: Set<string> = Array.from(ctf.values()).reduce(
    (prev, ctfNode) => {
      if (ctfNode.interfaces && ctfNode.interfaces.length) {
        ctfNode.interfaces.forEach(_interface => {
          const { subcommand } = _interface.module;
          prev.add(subcommand);
        });
      }
      return prev;
    },
    new Set()
  );

  stdSubCommands.forEach(stdSubCommand => {
    if (!ctfSubcommands.has(stdSubCommand)) {
      const stdCtfSkillToAdd = stdCtf.get(stdSubCommand);
      if (
        stdCtfSkillToAdd &&
        stdCtfSkillToAdd.interfaces &&
        stdCtfSkillToAdd.interfaces.length
      ) {
        stdCtfSkillToAdd.interfaces = normalizeInterfacesOfSkill(
          stdCtfSkillToAdd.interfaces
        );
      }
      ctf.set(stdCtfSkillToAdd.name, {
        ...stdCtfSkillToAdd,
        ...AddCtfHelpers
      });
    }
  });

  // Add all the CORE_CTF's without subcommands
  ctf.set('babel', { ...CORE_CTFS.babel, ...AddCtfHelpers });
  // @TODO
  // ctf.set('lodash', { ...CORE_CTFS.lodash, ...AddCtfHelpers });

  ctf.forEach(ctfNode => {
    const ctfWithHelpers = {
      ...ctfNode,
      ...AddCtfHelpers
    };
    Object.entries(ctfWithHelpers.ctfs || {}).forEach(([ctfName, ctfFn]) => {
      const correspondingCtfNode = ctf.get(ctfName);
      if (correspondingCtfNode) {
        ctf.set(
          ctfName,
          ctfFn(correspondingCtfNode, ctf, { alfredConfig, ...interfaceState })
        );
      }
    });
  });

  return ctf;
}

export async function loadConfigs(
  pkgPath: string = path.join(projectRoot, 'package.json')
): Promise<{ pkg: Object, pkgPath: string, alfredConfig: AlfredConfig }> {
  if (!fs.existsSync(pkgPath)) {
    throw new Error('Current working directory does not have "package.json"');
  }

  // Read the package.json and validate the Alfred config
  const pkg = JSON.parse((await fs.promises.readFile(pkgPath)).toString());
  const tmpAlfredConfig = pkg.alfred || {};
  ValidateConfig(tmpAlfredConfig || {});

  const defaultOpts = {
    npmClient: 'npm',
    skills: [],
    root: projectRoot
  };
  const alfredConfig = Object.assign({}, defaultOpts, tmpAlfredConfig);

  return { pkg, pkgPath, alfredConfig };
}

export default async function generateCtfFromConfig(
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): Promise<CtfMap> {
  // Check if any valid entrypoints exist
  const states = generateInterfaceStatesFromProject();
  if (!states.length) {
    throw new Error(
      `The project must have at least one entrypoint. Here are some examples of entrypoints:\n\n${ENTRYPOINTS.map(
        e => `"./src/${e}"`
      ).join('\n')}`
    );
  }

  // Generate the CTF
  const tmpCtf: CtfMap = new Map();
  const { skills = [] } = alfredConfig;
  skills.forEach(skill => {
    /* eslint-disable */
    const c = require(skill);
    /* eslint-enable */
    tmpCtf.set(c.name, c);
  });

  const ctf = CTF(Array.from(tmpCtf.values()), alfredConfig, interfaceState);
  addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState);

  return ctf;
}

export async function diffCtfDepsOfAllInterfaceStates(
  prevAlfredConfig: AlfredConfig,
  currAlfredConfig: AlfredConfig
): Set<string> {
  const stateWithDuplicateDeps = await Promise.all(
    INTERFACE_STATES.map(state =>
      Promise.all([
        generateCtfFromConfig(prevAlfredConfig, state),
        generateCtfFromConfig(currAlfredConfig, state)
      ])
    )
  );

  return Array.from(
    new Set(
      stateWithDuplicateDeps
        .map(([a, b]) => diffCtfDeps(a, b))
        .reduce((prev, curr) => prev.concat(curr), [])
    )
  );
}
