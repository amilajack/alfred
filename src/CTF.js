import lodash from 'lodash';

// @flow
type CtfNode = {
  name: string,
  interfaces: Array<string> | string,
  dependencies: {
    [x: string]: any
  },
  description: string,
  configFiles: Array<{
    name: string,
    path: string,
    config: {
      [x: string]: any
    }
  }>,
  ctfs: {
    [x: string]: (CtfNode, Map<string, CtfNode>) => CtfNode
  }
};

export const babel: CtfNode = {
  name: 'babel',
  description: 'Transpile JS from ESNext to the latest ES version',
  interfaces: 'alfred-interface-transpile',
  dependencies: {
    '@babel/cli': '7.2.0',
    '@babel/core': '7.2.0',
    '@babel/preset': 'env@7.2.0'
  },
  configFiles: [
    {
      name: 'babelrc',
      path: '.babelrc.js',
      config: {
        extends: '@babel/preset-env'
      }
    }
  ],
  ctfs: {
    webpack(config) {
      return config
        .extendConfig('webpack.base', {
          module: {
            devtool: 'source-map',
            mode: 'production',
            target: 'electron-main',
            entry: './app/main.dev',
            output: {
              path: 'app',
              filename: './app/main.prod.js'
            }
          }
        })
        .addDependencies({ 'babel-loader': '5.0.0' });
    },
    eslint(config) {
      return config
        .extendConfig('eslint', {
          parser: 'babel-eslint'
        })
        .addDependencies({ 'babel-eslint': '5.0.0' });
    }
  }
};

export const eslint: CtfNode = {
  name: 'eslint',
  description: 'Lint all your JS files',
  interfaces: 'alfred-interface-lint',
  dependencies: { eslint: '5.0.0' },
  configFiles: [
    {
      name: 'eslint',
      path: '.eslintrc.json',
      config: {
        extends: ['bliss']
      }
    }
  ],
  ctfs: {}
};

export const webpack: CtfNode = {
  name: 'webpack',
  description: 'Build, optimize, and bundle assets in your app',
  interfaces: 'alfred-interface-build',
  dependencies: { webpack: '5.0.0' },
  configFiles: [
    {
      name: 'webpack.base',
      path: 'webpack.base.js',
      config: {
        module: {
          rules: [
            {
              test: /\.jsx?$/,
              exclude: /node_modules/,
              use: {
                loader: 'babel-loader',
                options: {
                  cacheDirectory: true
                }
              }
            }
          ]
        },
        output: {
          path: "path.join(__dirname, '..', 'app')",
          // https://github.com/webpack/webpack/issues/1114
          libraryTarget: 'commonjs2'
        },
        resolve: {
          extensions: ['.js', '.jsx', '.json']
        },
        plugins: []
      }
    }
  ],
  ctfs: {}
};

type CtfHelpers = {
  findConfig: (configName: string) => { [x: string]: string },
  addDependencies: ({ [x: string]: string }) => { [x: string]: string },
  extendConfig: (x: string) => CtfNode
};

const AddCtfHelpers: CtfHelpers = {
  findConfig(configName: string) {
    return this.configFiles.find(configFile => configFile.name === configName);
  },
  extendConfig(
    configName: string,
    configExtension: { [x: string]: string } = {}
  ): CtfNode {
    const foundConfig = this.findConfig(configName);
    if (!foundConfig) {
      return this;
    }
    const mergedConfigFile = lodash.merge(foundConfig, {
      config: configExtension
    });
    const configFiles = this.configFiles.map(configFile =>
      configFile.name === configName ? mergedConfigFile : configFile
    );
    return lodash.merge(this, {
      configFiles
    });
  },
  addDependencies(dependencies) {
    return lodash.merge(this, {
      dependencies
    });
  }
};

export default function CTF(ctfs: Array<CtfNode>): Map<string, CtfNode> {
  const map: Map<string, CtfNode> = new Map();

  ctfs.forEach(_ctf => {
    const ctfWithHelpers = {
      ..._ctf,
      ...AddCtfHelpers
    };
    map.set(_ctf.name, ctfWithHelpers);
  });

  map.forEach(ctf => {
    Object.entries(ctf.ctfs || {}).forEach(([ctfName, ctfFn]) => {
      const correspondingCtfNode = map.get(ctfName);
      if (correspondingCtfNode) {
        map.set(ctfName, ctfFn(correspondingCtfNode, map));
      }
    });
  });

  return map;
}

// Intended to be used for testing purposes
export function getConfigs(
  ctf: Map<string, CtfNode>
): Array<{ [x: string]: any }> {
  return Array.from(ctf.values())
    .map(_ctf => _ctf.configFiles)
    .map(([config]) => config.config);
}

// Intended to be used for testing purposes
export function getDependencies(
  ctf: Map<string, CtfNode>
): { [x: string]: string } {
  return Array.from(ctf.values())
    .map(_ctf => _ctf.dependencies)
    .reduce((p, c) => ({ ...p, ...c }), {});
}
