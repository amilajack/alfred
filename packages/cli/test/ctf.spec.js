import path from 'path';
import CTF, { CORE_CTFS, INTERFACE_STATES, loadConfig } from '@alfredpkg/core';
import {
  diffCtfDeps,
  addMissingStdSkillsToCtf,
  diffCtfDepsOfAllInterfaceStates
} from '../src/helpers';
import generateCtfFromConfig from '../src/helpers/ctf';
import parcel from '../../alfred-skill-parcel';

const [defaultInterfaceState] = INTERFACE_STATES;

const defaultAlfredConfig = {
  root: path.join(__dirname, '../../../tests/fixtures/app'),
  skills: [],
  npmClient: 'npm'
};

describe('alfred cli helpers', () => {
  const { root: projectRoot } = defaultAlfredConfig;

  it('should install new deps after learning new skill', async () => {
    const { alfredConfig } = await loadConfig(projectRoot);
    const { webpack, babel } = CORE_CTFS;
    const oldCtf = CTF([webpack], alfredConfig, defaultInterfaceState);
    const newCtf = CTF([webpack, babel], alfredConfig, defaultInterfaceState);
    const ctf = diffCtfDeps(oldCtf, newCtf);
    expect(ctf).toMatchSnapshot();
  });

  it('should add missing std skills to ctf', async () => {
    {
      const { alfredConfig } = await loadConfig(projectRoot);
      const { webpack } = CORE_CTFS;
      const ctf = CTF([webpack], alfredConfig, defaultInterfaceState);
      expect(Array.from(ctf.keys())).toMatchSnapshot();
      expect(
        Array.from(
          addMissingStdSkillsToCtf(
            ctf,
            alfredConfig,
            defaultInterfaceState
          ).keys()
        )
      ).toMatchSnapshot();
    }
    {
      const { alfredConfig } = await loadConfig(projectRoot);
      const interfaceState = {
        env: 'production',
        projectType: 'app',
        target: 'browser'
      };
      const ctf = CTF([parcel], alfredConfig, interfaceState);
      expect(Array.from(ctf.keys())).toMatchSnapshot();
      const ctfSkillNames = Array.from(
        addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState).keys()
      );
      expect(ctfSkillNames).toMatchSnapshot();
      expect(ctfSkillNames).toContain('parcel');
      expect(ctfSkillNames).not.toContain('rollup');
      expect(ctfSkillNames).not.toContain('webpack');
    }
    {
      const { alfredConfig } = await loadConfig(projectRoot);
      const interfaceState = {
        env: 'production',
        projectType: 'lib',
        target: 'browser'
      };
      const ctf = CTF([parcel], alfredConfig, interfaceState);
      expect(Array.from(ctf.keys())).toMatchSnapshot();
      const ctfSkillNames = Array.from(
        addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState).keys()
      );
      expect(ctfSkillNames).toMatchSnapshot();
      expect(ctfSkillNames).toContain('rollup');
      expect(ctfSkillNames).not.toContain('parcel');
      expect(ctfSkillNames).not.toContain('webpack');
    }
  });

  describe('skills', () => {
    it('should throw if skill does not exist', async () => {
      const [state] = INTERFACE_STATES;
      const config = {
        ...defaultAlfredConfig,
        skills: [['@alfredpkg/skill-non-existent-skill', {}]]
      };
      await expect(generateCtfFromConfig(config, state)).rejects.toThrow(
        "Cannot find module '@alfredpkg/skill-non-existent-skill' from 'ctf.js'"
      );
    });

    it('should throw if unsupported skill is used', async () => {
      const [state] = INTERFACE_STATES;
      const config = {
        ...defaultAlfredConfig,
        skills: [['@alfredpkg/skill-non-existent-skill', {}]]
      };
      await expect(generateCtfFromConfig(config, state)).rejects.toThrow(
        "Cannot find module '@alfredpkg/skill-non-existent-skill' from 'ctf.js'"
      );
    });
  });

  it('should override core ctf skills that support same interface states', async () => {
    const { alfredConfig } = await loadConfig(projectRoot);
    const interfaceState = {
      env: 'production',
      projectType: 'app',
      target: 'browser'
    };
    const ctf = CTF([parcel], alfredConfig, interfaceState);

    expect(Array.from(ctf.keys())).toMatchSnapshot();
    const ctfSkillNames = Array.from(
      addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState).keys()
    );
    expect(ctfSkillNames).toMatchSnapshot();
    expect(ctfSkillNames).toContain('parcel');
    expect(ctfSkillNames).not.toContain('rollup');
    expect(ctfSkillNames).not.toContain('webpack');
  });

  it('should not use CTF skills that do not support current interface state', async () => {
    const { alfredConfig } = await loadConfig(projectRoot);
    const interfaceState = {
      env: 'production',
      projectType: 'lib',
      target: 'browser'
    };
    const ctf = CTF([parcel], alfredConfig, interfaceState);
    expect(Array.from(ctf.keys())).toEqual([]);
    const ctfSkillNames = Array.from(
      addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState).keys()
    );
    expect(ctfSkillNames).toMatchSnapshot();
    expect(ctfSkillNames).not.toContain('parcel');
    expect(ctfSkillNames).toContain('rollup');
    expect(ctfSkillNames).not.toContain('webpack');
  });

  it('should diff ctfs for all interface states', async () => {
    const skills = [
      ...defaultAlfredConfig.skills,
      '@alfredpkg/skill-parcel'
    ].map(e => [e, {}]);
    const currentAlfredConfig = {
      ...defaultAlfredConfig,
      skills
    };
    const result = await diffCtfDepsOfAllInterfaceStates(
      defaultAlfredConfig,
      currentAlfredConfig
    );
    expect(result).toEqual(['parcel@^1.11.0']);
  });
});
