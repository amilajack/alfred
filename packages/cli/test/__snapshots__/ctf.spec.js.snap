// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alfred cli helpers should add missing std skills to ctf 1`] = `
Array [
  "webpack",
]
`;

exports[`alfred cli helpers should add missing std skills to ctf 2`] = `
Array [
  "webpack",
  "eslint",
  "prettier",
  "jest",
  "babel",
]
`;

exports[`alfred cli helpers should add missing std skills to ctf 3`] = `
Array [
  "parcel",
]
`;

exports[`alfred cli helpers should add missing std skills to ctf 4`] = `
Array [
  "parcel",
  "eslint",
  "prettier",
  "jest",
  "babel",
]
`;

exports[`alfred cli helpers should add missing std skills to ctf 5`] = `Array []`;

exports[`alfred cli helpers should add missing std skills to ctf 6`] = `
Array [
  "eslint",
  "prettier",
  "rollup",
  "jest",
  "babel",
]
`;

exports[`alfred cli helpers should install new deps after learning new skill 1`] = `
Array [
  "babel-loader@5.0.0",
  "@babel/cli@7.2.0",
  "@babel/core@7.2.0",
  "@babel/preset-env@7.2.0",
]
`;

exports[`alfred cli helpers should not use CTF skills that do not support current interface state 1`] = `
Array [
  "eslint",
  "prettier",
  "rollup",
  "jest",
  "babel",
]
`;

exports[`alfred cli helpers should override core ctf skills that support same interface states 1`] = `
Array [
  "parcel",
]
`;

exports[`alfred cli helpers should override core ctf skills that support same interface states 2`] = `
Array [
  "parcel",
  "eslint",
  "prettier",
  "jest",
  "babel",
]
`;

exports[`alfred cli helpers skills should throw if skill does not exist 1`] = `[Error: Cannot find module '@alfredpkg/skill-non-existent-skill' from 'ctf.js']`;

exports[`alfred cli helpers skills should throw if unsupported skill is used 1`] = `[Error: The "mocha" skill, which supports {"env":["production","development","test"],"targets":["node"],"projectTypes":["app","lib"]}}, does not support the current environment, project type, or target, which are {"projectType":"app","target":"browser","env":"production"}]`;
